### PR TITLE
fix(host-service,desktop): stop the v1 import from claiming another repo's project via a secondary remote (#7241)

### LINKED ISSUE #7241: [bug] v1→v2 import hijacks another repo's v2 project via a non-origin remote, stranding the rightful repo's workspaces
## Summary

On a repo that has **more than one GitHub remote**, the v1→v2 import claimed the *wrong* cloud project: it matched a non-`origin` remote, hijacked the v2 project belonging to a **different** local repo, re-pointed that project's `repoPath` at the wrong directory, and left the rightful repo with no v2 project at all. The rightful repo's workspaces silently vanished from the sidebar.

App version 1.26.0, macOS.

## What I observed

`~/Kogan/kogan` (a monorepo, `origin` = `kogan/kogan`) has a second remote used for upstream syncs:

```
$ git -C ~/Kogan/kogan remote -v
oms-service  git@github.com:kogan/oms-service.git (fetch)
oms-service  git@github.com:kogan/oms-service.git (push)
origin       git@github.com:kogan/kogan.git (fetch)
origin       git@github.com:kogan/kogan.git (push)
```

`~/Kogan/oms-service` is a *separate* local repo (`origin` = `kogan/oms-service`) that was **already a v2 project** (`7efcc5af-…`, created 2026-05-11).

After the import ran:

- `v1_migration_state` recorded `01be230b (v1 "kogan", ~/Kogan/kogan) -> 7efcc5af` — i.e. the monorepo was imported **into the oms-service v2 project**.
- Cloud project `7efcc5af` was renamed `oms-service` → `kogan`, its repo flipped to `https://github.com/kogan/kogan`, and its path re-pointed to `~/Kogan/kogan`.
- The 8 monorepo workspaces were migrated onto `7efcc5af`.
- `~/Kogan/oms-service` had **no** `v1_migration_state` row at all, and none of its 7 v1 workspaces migrated. Its previously-existing v2 workspaces on `7efcc5af` were gone too.
- In the sidebar the monorepo appeared *labelled* `oms-service` with its own workspaces underneath, and the real oms-service project was absent. Re-adding `~/Kogan/kogan` as a project reported "already added".

Every worktree and branch survived on disk — this is a metadata/ownership bug, not disk loss — but all sidebar rows, names and workspace records for the displaced project were lost.

## Root cause

Three things compose into the failure. (Line refs from `dist/renderer/assets/layout-*.js` and `dist/main/host-service.js` in the 1.26.0 asar.)

**1. The importer opts into walking every remote.**

```js
function findProjectByPath(hostClient, project) {
  return hostClient.project.findByPath.query({
    repoPath: project.mainRepoPath,
    walkAllRemotes: true,
    expectedRemoteUrl: expectedRemoteUrlFor(project)
  });
}
```

`findByPath` then queries the cloud for **every** GitHub remote on the repo and merges the results:

```js
const allRemotes = await getGitHubRemotes(usage.createUserSimpleGit(gitRoot));
for (const parsed of allRemotes.values()) urlsToQuery.set(parsed.url.toLowerCase(), parsed);
...
const candidates = Array.from(byId.values()).sort((a, b) => {
  if (a.matchesExpected !== b.matchesExpected) return a.matchesExpected ? -1 : 1;
  return a.name.localeCompare(b.name);
});
```

`origin` gets no preference anywhere — not in the query set, not in the sort.

**2. The one disambiguation hint is dead in practice.**

```js
function expectedRemoteUrlFor(project) {
  if (!project.githubOwner) return void 0;
  ...
}
```

`projects.github_owner` is empty for **every** row in my v1 `local.db` (13/13). So `expectedRemoteUrl` is always `undefined`, `matchesExpected` is always `false`, and the sort silently degrades to `name.localeCompare` — an alphabetical tiebreak between unrelated repos.

**3. A lone wrong candidate is accepted without question.**

```js
function decideProjectImport(result) {
  const local = result.candidates.find((c) => c.source === "local-path");
  if (local) return { kind: "already-imported", v2ProjectId: local.id };
  if (result.candidates.length > 1) return { kind: "skip", reason: "multiple-candidates" };
  ...
  return { kind: "import" };
}
```

The `multiple-candidates` guard only fires at ≥2. In my case `kogan/kogan` had no v2 project yet, so exactly **one** candidate came back — the one found via the *non-origin* remote. `importV1Project` then takes `candidates[0]` blindly and calls `project.setup`:

```js
const targetCandidate = linkToProjectId ? candidates.find(c => c.id === linkToProjectId) : candidates[0];
...
await hostClient.project.setup.mutate({
  projectId: targetCandidate.id,
  origin: { repoCloneUrl: targetCandidate.repoCloneUrl, name: targetCandidate.name },
  mode: { kind: "import", repoPath: project.mainRepoPath, allowRelocate }
});
```

In `setup`'s `import` branch, `resolveMatchingSlug(repoPath, "kogan/oms-service")` happily finds the `oms-service` remote on the monorepo and returns the **monorepo** path, and `persistLocalProject` re-points the v2 project there. `rejectIfRepoint` doesn't help: it only guards when a local row for that project id already exists on this host.

Net: a repo that merely *references* another repo's remote can take that repo's v2 project.

## Repro

1. v2 project **A** exists for repo `owner/a`, checked out at `~/a`.
2. Separate v1 project **B** at `~/b`, `origin` = `owner/b`, plus a second remote pointing at `owner/a`.
3. No v2 project exists for `owner/b`.
4. Run the v1→v2 import.
5. **B** is imported into project **A**. **A** is renamed/re-pointed to `~/b`. `~/a` ends up with no v2 project and no migrated workspaces.

Multi-remote setups like this are ordinary: monorepo-to-service syncs, forks with an `upstream` remote, mirrors.

## Suggested fixes

- Prefer `origin` when building the candidate query set, and rank `origin`-derived candidates above other remotes in the sort.
- Don't accept a lone candidate that came from a non-`origin` remote without confirming with the user — treat "matched only via a secondary remote" the same way `multiple-candidates` is treated today.
- Never let `project.setup` re-point an existing cloud project's repo path/URL as an implicit side effect of import. If the resolved repo path differs from the project's recorded one, that's a conflict to surface, not a relocate.
- Populate/derive `githubOwner` (or just read `origin` directly) so `expectedRemoteUrl` isn't silently `undefined`; alternatively drop the hint and always resolve `origin` on the spot.
- Refuse to strand a v1 project: if another v1 project's import consumed the v2 project this one would have matched, record a `skip` with a reason rather than leaving no ledger row at all.

## Workaround (for anyone hit by this)

Worktrees and branches are untouched. Workspaces can be re-attached to the correct project — `workspaces create` adopts an existing worktree for an existing branch instead of creating a new one:

```
superset projects list          # find the correct project id
superset workspaces create --local --project <PROJECT_ID> \
  --name "<name>" --branch "<existing-branch>" --skip-branch-prefix
superset workspaces get <NEW_ID>   # confirm worktreePath is the pre-existing dir
```

Verified: `worktreePath` came back as the original worktree, `git worktree list` count unchanged, and uncommitted work in those worktrees survived.

### PR DESCRIPTION
**Pull request by Kitenite**
Fixes #7241.

## What

On a repo with more than one GitHub remote, the v1→v2 import could claim the v2 project of a *different* local repo: repo B (`origin` = owner/b, secondary remote → owner/a) was imported into project A, A's `repoPath` and repo URL were re-pointed at B's checkout, and A's own repo was left with no project. This PR closes all three holes that compose into that failure and adds a host-side backstop so `project.setup` can never move a project between repos.

## Why: the three holes

1. **Discovery walked every remote with no preference for `origin`.** `findByPath({walkAllRemotes: true})` queried the cloud for each GitHub remote and sorted candidates by `matchesExpected` then name, and the candidate shape carried no remote name, so the renderer could not tell an origin match from a secondary-remote match.
2. **The only disambiguation hint was dead.** `expectedRemoteUrlFor` depends on v1's `projects.github_owner`, which is empty in practice, so `matchesExpected` was always false and the sort degraded to alphabetical.
3. **A lone wrong candidate was accepted, and `setup` re-pointed the project.** `decideProjectImport` only skipped at two or more candidates; `importV1Project` took `candidates[0]` and called `project.setup {kind: "import"}`, whose `resolveMatchingSlug` matched the secondary remote and `persistLocalProject` overwrote `repoPath`/`repoOwner`/`repoName`/`repoUrl`. `rejectIfRepoint` only guards when a local row already exists, and the wizard's relocate confirm ("Use this folder") passes `allowRelocate: true` anyway.

## How

- **Host `findByPath`**: each cloud candidate gets `viaOrigin`; the result gets `hasOriginRemote` (true for an `origin` on any host, not just GitHub); origin-derived candidates sort first. The empty hint no longer decides anything.
- **Renderer `decideProjectImport`**: a lone candidate reached only through a secondary remote, on a repo that has an `origin`, is skipped with reason `non-origin-only` (both "Import all" and the headless boot pass). `importV1Project` never links to such a candidate implicitly and creates the folder as its own project; an explicit pick still links. Hosts that predate the flags keep today's behaviour.
- **Host `resolveMatchingSlug`** (used by `setup` import mode and the post-clone check): `CONFLICT` when the matching remote is not the folder's `origin` while an `origin` exists, regardless of `allowRelocate`. Relocating between two checkouts of the *same* repo still works. `findMatchingRemote` prefers `origin` when it matches.

The origin-ranking design (`viaOrigin` / `hasOriginRemote` / `non-origin-only`) is adopted from @andyst-dev's #7361, with thanks. This PR supersedes #7361: it carries that design plus the `setup` backstop the issue calls out as the third hole, the `importV1Project` guard, and the before/after reproduction.

## Proof

Reproduced and re-verified in the real dev desktop app driven over CDP, same harness both times: a real repo `a` (origin `fix7241/a`) as an existing host project with two workspaces and a cloud `v2_projects` row, plus a seeded v1 project `b` (origin `fix7241/b`, secondary remote `a` → `fix7241/a`, `github_owner` empty) with one v1 workspace.

Screenshots, DB dumps, and the exact step-by-step READMEs for both runs are on the verification page: https://app.superset.sh/page/pr-fix-7241-verified-v1-import-multi-remote-hijack-4ntvci (nothing binary is committed in this PR).

- **Before** (unmodified main): the wizard offers "Link" for `b`, "Import all" hits "Already set up at ~/tmp-7241/a. Link to ~/tmp-7241/b instead?", and "Use this folder" re-points project A's `repo_path` and its main workspace to `~/tmp-7241/b` with `remote_name = a`; B's workspace is adopted under A; the sidebar shows B's `feature` under `a`. DB dumps included.
- **After** (this branch): "Import all" no longer includes `b`; the headless pass ledgers `b` as `skipped / non-origin-only` with A untouched; the exact `setup` call the old prompt made is refused with `CONFLICT` with and without `allowRelocate` (`setup-guard.json`); the manual "Link" creates `b` as its own project, and the sidebar shows `a` and `b` side by side with their own workspaces.

## Tests

- `bun test apps/desktop/src/renderer/lib/v1-migration packages/host-service/src/trpc/router/project`: 109 pass. Seven new tests cover the three holes and fail on unmodified source (verified by reverting the four source files): `findByPath` flags a lone secondary-remote candidate and reports `hasOriginRemote` with no hint, ranks the origin candidate first, and handles a non-GitHub origin; `decideProjectImport` skips `non-origin-only`; `importV1Project` never auto-links such a candidate; `setup` import refuses a folder whose origin is another repo, with and without `allowRelocate`, and still relocates between checkouts of the same repo.
- `bun run typecheck` and `bun run lint:fix` clean. After the rebase onto main (#7480, #7514, #7516): 52 host-service and 96 desktop tests pass in the touched directories, both package typechecks pass, biome clean on touched files.

## Edge cases

`4943b8d555` adds 22 tests (170 pass across the two touched directories); the full matrix with expected outcomes is on the verification page under "Edge cases tested". In short:

- **Discovery**: origin-only; no remotes; secondary-only clones with no origin (still importable); non-GitHub and local-path origins; ssh/https and mixed-case forms of one repo collapsing to one origin-derived candidate; an `expectedRemoteUrl` hint pointing at the secondary's project (origin still ranks first); origin lookup failing while the secondary succeeds; the same project id resolved in either remote order keeps the origin URL.
- **Setup import**: origin mismatch rejected with and without `allowRelocate`, with and without a local row, never with the relocate-prompt phrase; secondary-only clones accepted; a duplicate secondary listed before origin still records `origin`; relocating between checkouts of the same repo still works.
- **Renderer**: hosts without the flags keep importing; `not-a-git-repo` and `multiple-candidates` keep precedence; a secondary-only candidate next to a failed origin lookup skips as `non-origin-only`; the host's origin-mismatch `CONFLICT` surfaces as an error, never as needs-relocate.
- **Headless pass**: the skipped project keeps the flip gate open, is re-evaluated without mutations on the next pass, and links once the folder is imported on its own, with its workspace adopted under the right project.

An independent second run (69 scratch tests written from the diff alone, plus path canonicalisation, `ssh://` form, a remote named `origin.url`, empty remote URLs) found no bug. One behaviour change to know about: a project that was already mis-linked before this fix now gets `CONFLICT` on a re-`setup` at its wrong path instead of a silent no-op; every `setup` caller is an explicit user action and relocating the project back to its real checkout still works, so this surfaces the corruption without blocking recovery.

## Review

CodeRabbit's finding at `project.ts:566` (a candidate first seen via a secondary remote and then upgraded by the origin lookup kept the secondary URL) is fixed in e9b77fbc85 with a regression test that queries the secondary remote first.

## Follow-up (out of scope here)

The wizard row for a `non-origin-only` project still labels its manual action "Link" although it now creates a fresh project; the label lives in the wizard's `import-plan.ts`.

https://claude.ai/code/session_015imdGkseBm3Lk7t5rycT4x

## Summary by CodeRabbit

- **Bug Fixes**
  - Project imports now prioritize candidates associated with the repository’s primary remote.
  - Projects found only through secondary remotes are no longer linked automatically and require explicit selection.
  - Prevented imports from linking to a different repository that merely references the expected project.
  - Valid checkouts of the same repository continue to support relocation and automatic linking.
  - Improved handling for repositories without a primary remote and for non-GitHub remotes.

### CODE REVIEW

File: packages/host-service/src/trpc/router/project/project.ts [hunk 1]
```diff
@@ -548,6 +563,7 @@ export const projectRouter = router({
 							// candidate and merge the match flag.
 							existing.matchesExpected =
 								existing.matchesExpected || matches(parsed.url);
+							existing.viaOrigin = existing.viaOrigin || viaOrigin;
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/host-service/src/trpc/router/project/project.ts]: Fixed in e9b77fbc85: when the origin lookup upgrades a candidate first seen through a secondary remote, the merge now replaces `repoCloneUrl` with the origin-derived URL instead of keeping the secondary one, so `project.setup` derives the origin slug. Regression test `keeps repoCloneUrl origin-derived when a secondary remote resolves the same project first` adds the remotes in `mirror` then `origin` config order and has both resolve to the same project id; it fails without the change and passes with it.
